charmcraft/commands/build.py

```diff
@@ -0,0 +1,284 @@
+
+    def __init__(self, args):
+        self.charmdir = args['from']
+        self.entrypoint = args['entrypoint']
+        self.requirement_paths = args['requirement']
+
+        self.buildpath = self.charmdir / BUILD_DIRNAME
+
+    def run(self):
+        """Main building process."""
+        logger.debug("Building charm in %r", str(self.buildpath))
+
+        if self.buildpath.exists():
+            shutil.rmtree(str(self.buildpath))
+        self.buildpath.mkdir()
+
+        linked_entrypoint = self.handle_code()
+        self.handle_dispatcher(linked_entrypoint)
+        self.handle_dependencies()
+        zipname = self.handle_package()
+
+        logger.info("Done, charm left in %r", zipname)
+        return zipname
+
+    def _link(self, srcpath, destdir=None):
+        """Link a file."""
+        if destdir is None:
+            destdir = self.buildpath
+        else:
+            if not destdir.exists():
+                destdir.mkdir()
+
+        destpath = destdir / srcpath.name
+        destpath.symlink_to(srcpath)
+        return destpath
+
+    def handle_code(self):
+        """Handle basic files and the charm source code."""
+        # basic files
+        logger.debug("Linking in basic files and charm code")
+        self._link(self.charmdir / CHARM_METADATA)
+
+        # the whole dir/tree if entry point is in a project's subdir, itself alone otherwise
+        if self.charmdir in self.entrypoint.parents and self.charmdir != self.entrypoint.parent:
+            # link the whole dir
+            linked_subdir = self._link(self.entrypoint.parent)
+            linked_entrypoint = linked_subdir / self.entrypoint.name
+        else:
+            # just the entry point, putting it under a src dir for juju compliance
+            linked_entrypoint = self._link(self.entrypoint, self.buildpath / 'src')
+        return linked_entrypoint
+
+    def handle_dispatcher(self, linked_entrypoint):
+        """Handle modern and classic dispatch mechanisms."""
+        # dispatch mechanism
+        current_dispatch = self.charmdir / DISPATCH_FILENAME
+        if current_dispatch.exists():
+            logger.debug("Including the current dispatch script")
+            dispatch_path = self._link(current_dispatch)
+        else:
+            logger.debug("Creating the dispatch mechanism")
+            dispatch_content = DISPATCH_CONTENT.format(
+                entrypoint_relative_path=linked_entrypoint.relative_to(self.buildpath))
+            dispatch_path = self.buildpath / DISPATCH_FILENAME
+            with dispatch_path.open("wt", encoding="utf8") as fh:
+                fh.write(dispatch_content)
+                fileno = fh.fileno()
+                os.fchmod(fileno, os.fstat(fileno).st_mode | stat.S_IXUSR)
+
+        # bunch of symlinks, to support old juju
+        current_hookpath = self.charmdir / 'hooks'
+        dest_hookpath = self.buildpath / 'hooks'
+        dest_hookpath.mkdir()
+        for hookname in HOOK_NAMES:
```

Comment:
I'm a bit surprised with this being "check to see if the charm declares a hooks/* object and then symlink it", but only do that for the 3 hooks we must have. It also isn't checking if 'hooks/install' is just a symlink to src/charm.py, and this would do 
in build/hooks/install => ../../hooks/install => ../src/charm.py

Reply:
As we talked on IRC:

1. You're right, I'm not checking "all the other stuff under hooks"... changed the code to "respect" everything the user has under hooks, and also verify that the "mandatory" ones are present, else create them

2. That symlink chain works for now, as the zip is putting just the content; but let's talk about this in the future when we revisit this "why all links?" -- as said above, please open an issue with your ideas